### PR TITLE
Migrate ipfs v0.7.0 to v0.8.0

### PR DESCRIPTION
**Pull request by gammazero**
- Convert from mdag pinning to datastore pinning

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -2,4 +2,16 @@ module github.com/ipfs/fs-repo-migrations
...
+	github.com/ipfs/go-ds-flatfs v0.4.5
+	github.com/ipfs/go-ds-leveldb v0.4.2
+	github.com/ipfs/go-filestore v1.0.0
+	github.com/ipfs/go-ipfs-blockstore v1.0.1
+	github.com/ipfs/go-ipfs-exchange-offline v0.0.1
+	github.com/ipfs/go-ipfs-pinner v0.0.5-0.20201029233447-cad837873ee6
+	github.com/ipfs/go-ipld-format v0.2.0
+	github.com/ipfs/go-merkledag v0.3.2
+	golang.org/x/net v0.0.0-20200114155413-6afb5195e5aa
+)
```
Comment (aschmahmann) [inline go.mod:17]: @Stebalien WDYT is this fine for now and then we move to implementing https://github.com/ipfs/fs-repo-migrations/issues/98, or should we just do that now?

Comment (aschmahmann) [reply go.mod:17]: Note: there might be alternatives to #98 such as vendoring the dependencies and giving them all migration-specific names so we can keep everything in one binary, but I'm not sure if they'd be any better.

Comment (Stebalien) [reply go.mod:17]: We can't have unvendored deps in this repo but we could try using go mod vendoring.

Comment (aschmahmann) [reply go.mod:17]: > The go mod vendor command constructs a directory named vendor in the main module's root directory that contains copies of all packages needed to support builds and tests of packages in the main module.

Is the suggestion to make new packages + go.mod's in each migration and then to try and vendor, or to copy the approach from the migrations before we were using gx.

File: ipfs-10-to-11/go.mod [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+require (
+	github.com/ipfs/fs-repo-migrations v1.6.4
+	github.com/ipfs/go-blockservice v0.1.2
+	github.com/ipfs/go-cid v0.0.7 // indirect
+	github.com/ipfs/go-datastore v0.4.5
+	github.com/ipfs/go-ds-flatfs v0.4.5
+	github.com/ipfs/go-ds-leveldb v0.4.2
+	github.com/ipfs/go-filestore v1.0.0
+	github.com/ipfs/go-ipfs-blockstore v1.0.0
```
Comment (aschmahmann) [inline ipfs-10-to-11/go.mod:13]: should use the pre v1 filestore + blockstore (same versions as in go-ipfs).

File: ipfs-10-to-11/migration/migration.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+			Prefix:    datastore.NewKey("/blocks"),
+			Datastore: fds,
+		},
+		{
+			Prefix:    datastore.NewKey("/"),
+			Datastore: ldb,
+		},
+	}))
+	var dstore datastore.Batching
+	dstore = &batchWrap{mdb}
```
Comment (aschmahmann) [inline ipfs-10-to-11/migration/migration.go:127]: This is specific to flatfs + leveldb, but people could be using anything. Instead use `go-ipfs/repo/fsrepo` with `repo, err := fsrepo.Open(repopath)` and `dstore = repo.Datastore()`

**Comment by Stebalien** [comment]
I've manually vendored the deps.

**Comment by gammazero** [comment]
One of the problems with using go modules for vendoring the dependencies of `ipfs-10-to-11`, and making `ipfs-10-to-11` into a module, is that building fs-repo-migrations requires the the `github.com/ipfs/fs-repo-migrations/ipfs-10-to-11` module.  Unfortunately, `ipfs-10-to-11` requires the `github.com/ipfs/fs-repo-migrations`.  This creates a circular dependency

I think the correct solution to this is presented in #98 which gets rid of the monolithic fs-repo-migrations binary.  As a temporary solution `ipfs-10-to-11` can depend on a previous version of `fs-repo-migrations` which had no changes to the packages imported, and did not depend on `ipfs-10-t0-11`.  This does not seem like a sustainable way of doing things, particularly if manual vendoring is required. (huge thanks to @Stebalien for doing that in this PR)

File: ipfs-10-to-11/migration/migration.go (hunk 3 above)
Comment (gammazero) [reply ipfs-10-to-11/migration/migration.go]: Done in [726f3c7](https://github.com/ipfs/fs-repo-migrations/pull/112/commits/726f3c787f63031cac157e40dfc4b653aa01f2a0)

**Comment by gammazero** [comment]
Note: Needed to update go version in top-level go.mod to `go 1.14` to fix "duplicate method" error when compiling go-ipfs for the 10-to-11 migration.  That error is caused by using the go 1.13 spec specified in go.mod.  See [here](https://dave.cheney.net/tag/interfaces) for details.

File: ipfs-10-to-11/go.mod (hunk 2 above)
Comment (gammazero) [reply ipfs-10-to-11/go.mod]: Done in [795e3b0](https://github.com/ipfs/fs-repo-migrations/pull/112/commits/795e3b0f85f0e01e10662ff64a050b533dd21724)

**Pull request merged**